### ISSUE TITLE
[Feature]: APIs traversing the trees using MySQL CTEs

### ISSUE THREAD

**Issue by dkarlovi** [pimcore/pimcore#12653]
### Feature description

In #11029 I suggested introducing additional guiding fields from the trees, using the nested set model. This is OK as a pragmatic solution since it's providing a good tradeoff between additional work when storing and performant reads. It does come with several drawbacks like needing to do it manually, it getting corrupted, etc.

What I missed there is Pimcore 10 requires MySQL 8+. This now means we have access to [CTEs](https://dev.mysql.com/doc/refman/8.0/en/with.html), the most interesting one is [`RECURSIVE` for hierarchical data traversal](https://dev.mysql.com/doc/refman/8.0/en/with.html#common-table-expressions-recursive-hierarchy-traversal).

TLDR try this:

```sql
WITH RECURSIVE `cte` (
    o_id,
    o_parentId,
    o_path,
    depth
) AS (
    SELECT
        o_id,
        o_parentId,
        CONCAT(o_path, o_key) AS o_path,
        1 AS `depth` 
    FROM `objects`
    WHERE `o_id` = :id
        UNION ALL 
    SELECT 
        o.o_id,
        o.o_parentId,
        CONCAT(o.o_path, o.o_key) AS o_path,
        `depth` + 1
    FROM `objects` o
    INNER JOIN cte
        ON o.`o_parentId` = cte.`o_id`
)
SELECT o_path FROM cte ORDER BY `depth` ASC;
```

On a 100k rows `objects` table it works *instantly*. You're able to plug in any ID and get it unroll all the children in some desired way. Going up the tree is just a matter of adjusting the join condition, like so:

```sql
    INNER JOIN cte
        ON o.`o_id` = cte.`o_parentId`
```

This can be used for any sort of tree traversal:
1. find all children
2. find children up to a certain depth
3. find children with additional conditions (for example, only published `Category` instances)
4. find parents (breadcrumbs), either toward root or away from root

The proposal here is to incorporate an CTE builder directly in the DAO in some way, which would cover at least these described use cases. This can then be used as a building block by components above it.

**Comment by brusch** [comment]
See also https://github.com/pimcore/pimcore/issues/9727#issuecomment-910634278

@markus-moser shared some feedback there, where it seems that the `LIKE` query on the `path` column is still faster.

**Comment by dkarlovi** [comment]
@brusch are all other things considered equal? For example, resource usage of the query (memory, CPU), etc? Like is also limited by length, you can't have really deep nested structures. Also, some features (like breadcrumbs) are not possible with LIKE.

It would be interesting to see the `EXPLAIN` for each query.

Edit: thinking about this further, some features which we should be getting with CTEs should be much slower with LIKE. For example, CTEs allows you to understand the structure, so if you need to unroll a branch depth-first or width first, you can do that. With LIKE I can't see how I'd do that in one query.

**Comment by dkarlovi** [comment]
@brusch short analysis here: https://github.com/pimcore/pimcore/issues/9727#issuecomment-1587424767

**Comment by fashxp** [comment]
won't implement that in admin UI classic anymore. in studio, we will use generic data index for improved performance.
